File: scripts/eu_evaporation_mc_boxplot.jl
# Monte-Carlo spread of the BEC atom number under RANDOM operational errors, for box plots.
# Unlike the worst-case (a min over a few one-axis extremes), here every axis is perturbed
# SIMULTANEOUSLY by a Gaussian draw of 1σ = ε on each of: power/α (common-mode depth), beam
# imbalance (H×(1+δ), V×(1−δ)), T₀, N₀. We draw M realizations at each ε and record N_BEC for
# BOTH the lab ramp and the (ε-invariant) optimum, so a box/whisker plot shows the real
# shot-to-shot distribution — the "many points" view. The optimum ramp is computed once.
#
# Usage: julia --project=. scripts/eu_evaporation_mc_boxplot.jl
using SpinorBEC, DelimitedFiles, Random
using SpinorBEC: Eu151, EvapParams, euv3_evap_trap, euv3_evaporation_ramp, euv3_defaults,
    FortRamp, run_evaporation, optimize_ramp_monotone, _EUV3_ALPHA

const OUT = joinpath(@__DIR__, "..", "docs", "guides", "figures")
const LEVELS = [0.0, 0.01, 0.02, 0.03, 0.05, 0.07, 0.10]   # 1σ operational-error levels
const NDRAW = 400

d = euv3_defaults()
trap = euv3_evap_trap()
p = EvapParams(; a_s=Eu151.a_s, tau_bg=d.tau_bg, K3=d.K3)
lab = euv3_evaporation_ramp()
N0, T0 = d.N0, d.T0

println("optimizing the (ε-invariant) nominal ramp…")
flush(stdout)
opt = optimize_ramp_monotone(trap, p, lab; N0=N0, T0=T0, restarts=4).ramp

# N_BEC for `ramp` under one operational-error realization (δp,δi,δT,δN)
function draw_nbec(ramp::FortRamp, δp, δi, δT, δN)
    t = euv3_evap_trap(; alpha=_EUV3_ALPHA * (1 + δp))       # common-mode depth ≡ α (αP degeneracy)
    pw = copy(ramp.powers_W)
    pw[1, :] .*= (1 + δi)                                    # H beam
    size(pw, 1) >= 2 && (pw[2, :] .*= (1 - δi))              # V beam (opposite ⇒ imbalance)
    r2 = FortRamp(ramp.times, pw)
    rr = run_evaporation(t, r2, p; N0=N0 * (1 + δN), T0=T0 * (1 + δT))
    rr.reached_bec ? rr.N_BEC : 0.0
end

rng = MersenneTwister(20260719)
rows = Vector{Any}[]
for ε in LEVELS
    for _ in 1:NDRAW
        δp, δi, δT, δN = ε .* randn(rng, 4)
        push!(rows, [Int(round(100ε)), draw_nbec(lab, δp, δi, δT, δN),
            draw_nbec(opt, δp, δi, δT, δN)])
    end
end
M = permutedims(hcat(rows...))
open(joinpath(OUT, "eu_evap_mc_boxplot.csv"), "w") do io
    println(io, "eps_pct,lab_NBEC,opt_NBEC")
    writedlm(io, M, ',')
end

# quick text summary (median + 5/95 pct of the optimum, and BEC-miss fraction)
println("\nε%   opt median   opt 5–95%          miss(opt)")
for ε in LEVELS
    v = [r[3] for r in rows if r[1] == Int(round(100ε))]
    sv = sort(v)
    q(f) = sv[clamp(round(Int, f * length(sv)), 1, length(sv))]
    miss = count(==(0.0), v) / length(v)
    println("  ", lpad(Int(round(100ε)), 2), "   ", lpad(round(Int, q(0.5)), 8),
        "   ", lpad(round(Int, q(0.05)), 7), "–", rpad(round(Int, q(0.95)), 7),
        "   ", round(100miss, digits=1), "%")
end
println("\nwrote ", joinpath(OUT, "eu_evap_mc_boxplot.csv"), "  (", size(M, 1), " draws)")

File: scripts/eu_evaporation_mc_boxplot.py
#!/usr/bin/env python3
"""Box-and-whisker of the ¹⁵¹Eu BEC atom number under random operational errors.

At each operational-error level ε (1σ, applied simultaneously to power/α, beam imbalance, T₀, N₀)
we Monte-Carlo NDRAW realizations and box-plot the resulting N_BEC for the optimum ramp (and the
lab ramp for reference). Individual draws are overlaid as jittered points. Shows the true
shot-to-shot spread, not just the worst case.

Run scripts/eu_evaporation_mc_boxplot.jl first to emit the CSV.
"""
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

FIG = Path(__file__).resolve().parent.parent / "docs" / "guides" / "figures"
d = np.loadtxt(FIG / "eu_evap_mc_boxplot.csv", delimiter=",", skiprows=1, ndmin=2)
eps_all = d[:, 0]
lab_all = d[:, 1] / 1e4
opt_all = d[:, 2] / 1e4
levels = np.unique(eps_all)

opt_by = [opt_all[eps_all == e] for e in levels]
lab_by = [lab_all[eps_all == e] for e in levels]
lab_med = np.median(lab_by[0])

fig, ax = plt.subplots(figsize=(9.5, 5.2))
pos = np.arange(len(levels))
w = 0.34

def box(data, positions, color, label):
    bp = ax.boxplot(data, positions=positions, widths=w, patch_artist=True,
                    whis=(5, 95), showfliers=False, manage_ticks=False)
    for b in bp["boxes"]:
        b.set(facecolor=color, alpha=0.35, edgecolor=color)
    for k in ("whiskers", "caps", "medians"):
        for x in bp[k]:
            x.set(color=color, linewidth=1.5 if k == "medians" else 1.0)
    bp["boxes"][0].set_label(label)
    return bp

box(opt_by, pos - w / 1.7, "tab:blue", "optimum ramp")
box(lab_by, pos + w / 1.7, "0.5", "lab ramp")

# jittered individual draws (the "many points")
rng = np.random.default_rng(0)
for i, (o, l) in enumerate(zip(opt_by, lab_by)):
    ax.scatter(pos[i] - w / 1.7 + rng.uniform(-0.06, 0.06, o.size), o,
               s=4, color="tab:blue", alpha=0.15, zorder=1)
    ax.scatter(pos[i] + w / 1.7 + rng.uniform(-0.06, 0.06, l.size), l,
               s=4, color="0.4", alpha=0.15, zorder=1)

ax.axhline(lab_med, color="0.5", ls=":", lw=1.2, label="lab ramp median (ε=0)")
ax.set_xticks(pos)
ax.set_xticklabels([f"{int(e)}" for e in levels])
ax.set_xlabel("operational-error level  ε  (% 1σ, all axes: power/α · imbalance · T₀ · N₀)")
ax.set_ylabel(r"BEC atom number  $N_{\rm BEC}\ (\times 10^4)$")
ax.set_title("BEC atom-number spread under random operational errors (box: 25–75%, whisker: 5–95%)")
ax.set_ylim(bottom=0)
ax.legend(fontsize=9, loc="lower left")
ax.grid(alpha=0.3, axis="y")

fig.tight_layout()
out = FIG / "eu_evaporation_mc_boxplot.png"
fig.savefig(out, dpi=150)
print("wrote", out)
